README.md
# hamt_py

Initial commit.
## On-line Documentation

More information on the **hamt_py** project can be found
[here](https://jddixon.github.io/hamt_py)


CHANGES
py/hamt_py/CHANGES

v0.1.17
    2018-03-23
        * fix corrupt .dvcz/builds                                  SLOC 1183
v0.1.16
    2018-03-02
        * edit setup.py, tox config files                           SLOC 1180
v0.1.15
    2017-12-15
        * pylinting, other code cleanup                             SLOC 1184
v0.1.14
    2017-10-02
        * add projConfig.toml                                       SLOC 1191
v0.1.13
    2017-09-15
        * pep8 etc; correct tox.ini, requirements.txt               SLOC 1159
v0.1.12
    2017-08-15
        * add tox.ini and *requirements.txt; edit cfg files         SLOC 1152
v0.1.11
    2017-07-16
        * add MANIFEST.in, update setup.py and cfg files            SLOC 1117
v0.1.10
    2017-06-10
        * restructure for pytest                                    SLOC 1107
v0.1.9
    2017-04-24
        * drop dist/* from repo; plus minor edits                   SLOC 1106
v0.1.8
    2017-03-17
        * complete and test {Table,Root}.delete_leaf()              SLOC 1103
v0.1.7
    2017-03-15
        * clean up, pylint code; fiddle with tests                  SLOC 1072
v0.1.6
    2017-03-14
        * add test_with_matching_keys: all tests succeed            SLOC 1072
v0.1.5
    2017-03-13
        * simple tets_table.py succeeds                             SLOC 1015
v0.1.4
    2017-03-12
        * improved test_root, less rough test_table                 SLOC 1007
v0.1.3
    2017-03-11
        * improved test_root, rough test_table                      SLOC  989
v0.1.2
    2017-03-10
        * test_root.py succeeds                                     SLOC  905
v0.1.1
    2017-03-08
        * skeletal test_root.py                                     SLOC  809
v0.1.0
    2017-03-07
        * complete Table                                            SLOC  693
v0.0.3
    2017-03-06
        * add Root and most of Table                                SLOC  603
v0.0.3
    2017-03-03
        * add pylint, coverage, Leaf, and test                      SLOC  386
v0.0.2
    2017-03-02
        * add configuration files
v0.0.1
    2017-03-02
        * initial commit
























test_leaf.py
#!/usr/bin/env python3
# hamt_py/test_basics.py

""" Test properties of the HAMT leaf. """

#import hashlib
#import os
import time
import unittest

from rnglib import SimpleRNG
from hamt import Leaf, HamtError


class TestLeaf(unittest.TestCase):
    """ Test properties of the HAMT leaf. """

    def setUp(self):
        self.rng = SimpleRNG(time.time())

    def tearDown(self):
        pass

    def test_ctor(self):
        """
        Test constructor functionality.
        """

        # tests that should raise an exception

        with self.assertRaises(HamtError):
            leaf = Leaf(None, 'some random value')
        with self.assertRaises(HamtError):
            leaf = Leaf(b'some key', None)

        # tests that should succeed
        key = self.rng.some_bytes(8)
        value = self.rng.some_bytes(16)
        leaf = Leaf(key, value)
        self.assertEqual(leaf.key, key)
        self.assertEqual(leaf.value, value)

        self.assertTrue(isinstance(leaf, Leaf))

if __name__ == '__main__':
    unittest.main()


test_root.py
#!/usr/bin/env python3
# hamt_py/test_root.py

""" Test properties of the HAMT Root. """

#import hashlib
#import os
import time
import unittest

from rnglib import SimpleRNG
from hamt import HamtNotFound, Root, Leaf, uhash, countem


class TestRoot(unittest.TestCase):
    """ Test properties of the HAMT Root. """

    def setUp(self):
        self.rng = SimpleRNG(time.time())

    def tearDown(self):
        pass

    MAX_T = 8

    def make_a_leaf_cluster(self, texp):
        """
        Make 1 << texp leafs with distinct keys and values.
        """
        if texp > TestRoot.MAX_T:
            raise RuntimeError(
                "THIS WON'T WORK: texp is %d but may not exceed MAX_T=%d" % (
                    texp, TestRoot.MAX_T))

        key = self.rng.some_bytes(8)
        value = self.rng.some_bytes(16)

        leaves = []
        count = 1 << texp
        mask = count - 1       # we want to mask off that many bits
        shift = 8 - texp       # number of bits to shift the mask
        if texp < 8:
            mask <<= shift

        # DEBUG
        # print("make_a_leaf_cluster: texp %d, shift %d, mask 0x%x" % (
        #    texp, shift, mask))
        # END

        key = self.rng.some_bytes(8)
        value = self.rng.some_bytes(16)

        for ndx in range(count):
            slot_nbr = ndx << shift

            mykey = bytearray(8)
            mykey[:] = key              # local copy of key
            myval = bytearray(16)
            myval[:] = value            # local copy of value
            mykey[0] &= ~mask
            mykey[0] |= slot_nbr
            # pylint: disable=redefined-variable-type
            mykey = bytes(mykey)

            myval[0] &= ~mask
            myval[0] |= slot_nbr
            myval = bytes(myval)

            leaf = Leaf(mykey, myval)
            leaves.append(leaf)
            # DEBUG
            # print("%3d key %s value %s" % (
            #    ndx, b2a_hex(mykey), b2a_hex(myval)))
            # END
        return leaves

    # ---------------------------------------------------------------

    def do_test_root_ctor(self, wexp):
        """ Test Root constructor with specific parameters. """

        texp = wexp + self.rng.next_int16(4)
        root = Root(wexp, texp)
        self.assertIsNotNone(root)
        self.assertEqual(root.wexp, wexp)
        self.assertEqual(root.texp, texp)
        self.assertEqual(root.max_table_depth, (64 - texp) // wexp)

        self.assertIsNotNone(root.slots)
        self.assertEqual(len(root.slots), 1 << texp)     # 0)

        self.assertEqual(root.slot_count, 1 << texp)
        self.assertEqual(root.mask, root.slot_count - 1)

    def test_ctor(self):
        """
        Test constructor functionality with a range of parameters.
        """
        for wexp in [3, 4, 5, 6]:
            self.do_test_root_ctor(wexp)

    # ---------------------------------------------------------------

    def do_test_flat_root(self, wexp):
        """ Test insertions only into the root table for the value of wexp. """

        texp = wexp     # we aren't interested in texp != wexp
        root = Root(wexp, texp)
        self.assertEqual(root.leaf_count, 0)
        self.assertEqual(root.table_count, 1)     # root table is counted

        candidate = self.make_a_leaf_cluster(texp)     # 1 << texp unique Leafs
        inserted = 0
        ndxes = []
        leaves = []
        for leaf in candidate:
            ndx = uhash(leaf.key) & root.mask
            if not ndx in ndxes:
                root.insert_leaf(leaf)
                leaves.append(leaf)
                inserted += 1
                ndxes.append(ndx)
                self.assertEqual(root.leaf_count, inserted)
                self.assertEqual(root.table_count, 1)
                value = root.find_leaf(leaf.key)
                self.assertEqual(value, leaf.value)

        # EXPERIMENT
        # print("\nroot.leaf_count is %d" %  root.leaf_count)
        # root.visit(countem)

        for leaf in leaves:

            # verify the slot is occupied -----------------
            hcode = uhash(leaf.key)
            ndx = root._mask & hcode
            self.assertIsNotNone(root._slots[ndx])      # SOMETIMES FAILS

            # verify the leaf is present ------------------
            found = root.find_leaf(leaf.key)
            self.assertIsNotNone(found)

            # ... and its value is as expected ------------
            self.assertEqual(found, leaf.value)

            # remove the leaf -----------------------------
            root.delete_leaf(leaf.key)
            # slot is empty -------------------------------
            self.assertIsNone(root._slots[ndx])
            # the leaf is gone ----------------------------
            try:
                value = root.find_leaf(leaf.key)
            except HamtNotFound:
                # success
                pass

        self.assertEqual(root.leaf_count, 0)
        self.assertEqual(root.table_count, 1)

    def test_flat_root(self):
        """
        Test constructor functionality with a range of parameters.
        """
        for wexp in [3, 4, 5, 6]:
            self.do_test_flat_root(wexp)


if __name__ == '__main__':
    unittest.main()


src/hamt/__init__.py
# hamt/__init__.pyMT

""" NodeID library for python XLattice packages. """

import sys
# from binascii import b2a_hex

from xlutil import popcount64

__version__ = '0.1.17'
__version_date__ = '2018-03-23'

__all__ = ['__version__', '__version_date__',
           'MAX_W',
           'countem',       # EXPERIMENT
           'uhash',
           'HamtError', 'HamtNotFound',
           'Leaf', 'Table', 'Root']

# CONSTANTS

MAX_W = 6

# FUNCTIONS


def uhash(val):
    """ Return the hash of a string of bytes as an unsigned int64. """
    return hash(val) % ((sys.maxsize + 1) * 2)

# EXPERIMENT --------------------------------------------------------


def countem(self):
    """ Count the number of Leafs and Tables in a table. """

    # pylint: disable=protected-access
    table_count, leaf_count = 0, 0
    for slot in self._slots:
        if slot:
            if isinstance(slot, Table):
                table_count += 1
            elif isinstance(slot, Leaf):
                leaf_count += 1
    # DEBUG
    print("countem: %d tables, %d leafs" % (table_count, leaf_count))
    # END

# END EXPERIMENT ----------------------------------------------------

# CLASSES


class HamtError(RuntimeError):
    """ General purpose exception for the package. """


class HamtNotFound(HamtError):
    """ Raised where there is no match for a key. """


class Leaf(object):
    """The Leaf in a HAMT data structure. """

    def __init__(self, key, value):

        if key is None:
            raise HamtError('key cannot be None')
        if value is None:
            raise HamtError("leaf value cannot be none")
        self._key = key
        self._value = value

    @property
    def key(self):
        """ Return the key part of a HAMT Leaf."""
        return self._key

    @property
    def value(self):
        """ Return the value pointed at by a HAMT Leaf. """
        return self._value


class Table(object):
    """
    Table with a dynamic number of entries.

    The number of slots may grow to to a maximum of (1 << wexp).
    Each slot is either empty or points to either a Leaf or another Table.

    Unlike the Go version, a hamt_py Table can only be created if its
    first leaf is specified.  That is, __init__() below is equivalent to
    Go hamt_go's NewTableWithLeaf.
    """

    last_nbr = -1

    @staticmethod
    def check_table_param(depth, root):  # -> (int, int)
        """ Raise if table parameter is out of range. """
        if root is None:
            raise HamtError("root must have a value")
        wexp = root.wexp
        texp = root.texp
        if wexp > MAX_W:
            raise HamtError("Max table size exceeded.")
        if texp + (depth - 1) * wexp > 64:
            raise HamtError("Max table depth exceeded.")
        return wexp, texp           # Seen as unsigned ints

    def __init__(self, depth, root, first_leaf):

        Table.last_nbr += 1
        self._nbr = Table.last_nbr

        # set up the table
        self._depth = depth
        wexp, texp = Table.check_table_param(depth, root)
        self._wexp = wexp
        self._texp = texp
        self._root = root           # so we can determine max_table_depth
        wflag = 1 << wexp           # seen as uint64
        self._mask = wflag - 1      # ditto

        # insert the first leaf
        shift_count = texp + (depth - 1) * wexp
        hcode = uhash(first_leaf.key) >> shift_count
        ndx = hcode & self._mask    # index into bit map
        flag = 1 << ndx             # seen as uint64
        self._slots = [first_leaf]
        self._bitmap = flag         # set bit for this entry

        # DEBUG
#       print("new Table[%d]:  depth  = %d" % (self._nbr, depth))
#       print("                texp   = %d" % texp)
#       print("                wexp   = %d" % wexp)
#       print("                wflag  = %d (0x%x)" % (wflag, wflag))
#       print("                mask   = 0x%x" % self._mask)
#       print("    first leaf: hcode  = 0x%x" % hcode)
#       print("                ndx    = %d (0x%x)" % (ndx, ndx))
#       print("                flag   = %d (0x%x)" % (flag, flag))
#       print("    Table[%d]:  bitmap = 0x%x" % (self._nbr, self._bitmap))
#       print()
#       # END

    @property
    def root(self):
        """ Return the Root of the table. """
        return self._root

    @property
    def wexp(self):
        """
        Return the w factor, where 2^w is the number of slots in the Table.
        """
        return self._wexp

    @property
    def texp(self):
        """
        Return the t factor, where 2^t is the number of slots in the Root.
        """
        return self._texp

    @property
    def mask(self):
        """ Return a bit vectory with the bit set if the slot is free. """
        return self._mask

    @property
    def slots(self):
        """ Return a list of slots in use. """
        return self._slots

    @property
    def bitmap(self):
        """ Return a bitmap with the Nth bit set to 1 if slot N is in use. """
        return self._bitmap

    @property
    def max_slots(self):
        """
        Return one more than the maximum number of slots that may be
        occupied.
        """
        return 1 << self._wexp

    @property
    def leaf_count(self):
        """ Return a count of the leaf nodes in this Table, recursing. """
        count = 0
        # for node in self._slots:
        for node in self._slots:
            if node:
                if isinstance(node, Leaf):
                    count += 1
                elif isinstance(node, Table):
                    count += node.leaf_count
        # DEBUG
#       print("Table[%d].leaf_count, depth %d, returning %d" % (
#           self._nbr, self._depth, count))
        # END
        return count

    @property
    def table_count(self):
        """
        Return a count of Table nodes below this Table, including this Table.
        """
        count = 1       # this Table
        for node in self._slots:
            if node and isinstance(node, Table):
                count += node.table_count
        return count

    def remove_from_slots(self, offset):
        """ Remove an entry from this Table. """

        cursize = len(self._slots)
        if cursize == 0:
            raise HamtError("attempt to delete from empty Table")
        if offset >= cursize:
            msg = 'Internal error: delete offset %d but table size %d' % (
                offset, cursize)
            raise HamtError(msg)
        elif cursize == 1:
            # LEAVES EMPTY TABLE, WHICH WILL HARM PERFORMANCE
            self._slots = []
        elif offset == 0:
            self._slots = self._slots[1:]
        elif offset == cursize - 1:
            self._slots = self._slots[:-1]
        else:
            self._slots = self._slots[:offset] + self._slots[offset + 1:]

    def delete_leaf(self, hcode, key):
        """
        Remove a Leaf from the Table.

        Enter with hcode the hashcode for the key shifted appropriately
        for the current depth, so that the first w bits of the shifted
        hashcod can be used as the index of the leaf in the table.

        The caller guarantees that depth <= root.max_table_depth.
        """

        if not self._slots:
            raise HamtNotFound

        ndx = hcode & self._mask
        flag = 1 << ndx             # a uint64
        mask = flag - 1
        if self._bitmap & flag == 0:
            raise HamtNotFound

        # the node is present: get its position in the slice
        slot_nbr = 0
        if mask:
            slot_nbr = popcount64(self._bitmap & mask)
        node = self._slots[slot_nbr]
        if isinstance(node, Leaf):
            # key = node.key        # REDUNDANT?
            if node.key == key:
                self.remove_from_slots(slot_nbr)
                self._bitmap &= ~flag           # TEST
            else:
                raise HamtNotFound
        else:
            # node is a table, so recurse
            if self._depth + 1 > self.root.max_table_depth:
                raise HamtNotFound
            node.delete_leaf(hcode >> self._wexp, key)

    def find_leaf(self, hcode, depth, key):
        """
        Find a Leaf in the Table using the hashcode hcode and matching
        depth and key; return the value of the Leaf.

        The hashcode has been shifted appropriately for the current
        depth (zero-based).  Return None if no matching entry is found
        or the entry associated with the key.

        The caller guarantees that depth <= Root.max_table.depth.
        """

        value = None
        slot_nbr = 0
        ndx = hcode & self._mask    # uint
        flag = 1 << ndx             # uint64

        # DEBUG
#       print("Table[%d].find_leaf:" % self._nbr)
#       print("    depth    %d" % depth)
#       print("    hcode    0x%x" % hcode)
#       print("    mask     0x%x" % self._mask)
#       print("    ndx      %d (0x%s)" % (ndx, ndx))
#       print("    flag     0x%x " % flag)
#       print("    bitmap   0x%x " % self._bitmap)
        # END

        if self._bitmap & flag:
            # the node is present
            mask = flag - 1         # also uint64
            if mask:
                slot_nbr = popcount64(self._bitmap & mask)
            # DEBUG
#           print("    slot_nbr 0x%x " % slot_nbr)
            # END
            node = self._slots[slot_nbr]
            if isinstance(node, Leaf):
                if key == node.key:
                    value = node.value
                # DEBUG
                else:
                    print("Table[%d].find_leaf: key mismatch, returning None" %
                          self._nbr)
                # END
            else:
                # node is a Table, so recurse
                if depth <= self.root.max_table_depth:
                    hcode >>= self._wexp
                    value = node.find_leaf(hcode, depth + 1, key)
                    # DEBUG
#                   if value is None:
#                       print("Table[%d].find_leaf: recursing returned None" %
#                             self._nbr)
                    # END
                # otherwise we will return None
        # DEBUG
#       else:
#           print("Table[%d].find_leaf:  depth %d: bitmap is empty" % (
#               self._nbr, depth))
        # END
        return value    # we already have the key

    def insert_leaf(self, hcode, leaf):
        """
        Enter with hcode having been shifted so that the low-order wexp bits
        determine ndx, the index of the bit to be set`.

        The caller guarantees that depth <= self.root.max_table_depth.
        """

        slot_nbr = 0
        ndx = hcode & self._mask    # mask off wdx low-order bits
        flag = 1 << ndx             # uint64
        mask = flag - 1
        if mask:
            # count bits below this one in the bitmap
            slot_nbr = popcount64(self._bitmap & mask)
        slice_size = len(self._slots)
#       # DEBUG
#       print("Table[%d].insert_leaf: PRE-INSERTION" % self._nbr)
#       print("                   depth      = %d"   % self._depth)
#       print("                   hcode      = 0x%x" % hcode)
#       print("                   ndx        = %d (0x%x)" % (ndx, ndx))
#       print("                   flag       = 0x%x" % flag)
#       print("                   mask       = 0x%x" % mask)
#       print("                   slot_nbr   = %d" % slot_nbr)
#       print("                   bitmap     = 0x%x" % self._bitmap)
#       print("                   slice_size = %d" % slice_size)
#       # END

        if slice_size == 0:
            self._slots = [leaf]
            self._bitmap |= flag
        else:
            # is there already something in this slot?
            if self._bitmap & flag:
                entry = self._slots[slot_nbr]
                if entry is None:
                    print("INTERNAL ERROR: flag=%d, slot=%d is empty" % (
                        flag, slot_nbr))

                if isinstance(entry, Leaf):
                    if entry.key == leaf.key:
                        # keys match so replace value
                        entry.value = leaf.value        # MUST BE DEEPCOPY ?
                    else:
                        deeper = Table(self._depth + 1, self._root, entry)
                        deeper.insert_leaf(hcode >> self.wexp, leaf)

                        # the new table replaces the existing leaf
                        self._slots[slot_nbr] = deeper

                else:
                    # it's a Table, so let's recurse
                    entry.insert_leaf(hcode >> self._wexp, leaf)

            # nothing in the slot
            elif slot_nbr == 0:
                self._slots = [leaf] + self._slots
                self._bitmap |= flag
            elif slot_nbr == slice_size:
                self._slots += [leaf]
                self._bitmap |= flag
            else:
                self._slots = self._slots[:slot_nbr] + \
                    [leaf] + self._slots[slot_nbr:]
                self._bitmap |= flag

        # DEBUG
#       print("Table[%d].insert_leaf: POST INSERTION depth %d, bitmap 0x%x" % (
#           self._nbr, self._depth, self._bitmap))
        # END


class Root(object):
    """
    Root table of a HAMT Trie.

    The Root has a fixed number of slots, each of which may be empty
    or may point to a Table or a Leaf.  There are (1 << texp) slots
    in the Root table.
    """

    def __init__(self, wexp, texp):
        if wexp < 2:
            raise HamtError("w cannot be less than 2, is %d" % wexp)
        if texp < 2:
            raise HamtError("w cannot be less than 2, is %d" % texp)
        if wexp > MAX_W:
            raise HamtError("max table size (%d) exceeded" % MAX_W)
        if texp > 64:
            raise HamtError("max root table size (64) exceeded")
        flag = 1 << texp    # number of slots available

        self._wexp = wexp
        self._texp = texp
        self._max_table_depth = (64 - texp) // wexp
        self._slot_count = flag
        self._mask = flag - 1
        self._slots = [None] * flag
        # DEBUG
        # print("Root: wexp            %d" % wexp)
        # print("      texp            %d" % texp)
        # print("      max table depth %d" % self.max_table_depth)
        # print("      slot count      %d" % flag)
        # print("      mask            0x%x" % self._mask)
        # END

    @property
    def wexp(self):
        """
        Return the w factor, where 2^w is the number of slots in the Table.
        """
        return self._wexp

    @property
    def texp(self):
        """
        Return the t factor, where 2^t is the number of slots in the Root.
        """
        return self._texp

    @property
    def max_table_depth(self):
        """
        Return the nmaximum number of slots in the Root or any descendents.
        """
        return self._max_table_depth

    @property
    def slot_count(self):
        """ Return the number of slots in the root table.  """
        return self._slot_count

    @property
    def mask(self):
        """ A mask of bits: 1 if the slot is free, 0 if not. """
        return self._mask

    @property
    def slots(self):
        """ Return the slots table for the Root. """
        return self._slots

    @property
    def leaf_count(self):
        """ Return a count of leaf nodes under the Root. """
        count = 0
        # for node in self._slots:
        for node in self._slots:
            if node:
                if isinstance(node, Leaf):
                    count += 1
                    # DEBUG
                    # print("root slot %3d is Leaf" % ndx)
                    # END
                elif isinstance(node, Table):
                    count += node.leaf_count
        # DEBUG
            # else:
            #    print("root slot %03d is EMPTY" % ndx)
        # print("             returning Root.leaf_count %d" % count)
        # END
        return count

    @property
    def table_count(self):
        """
        Return a count of Tables under the Root, including the Root itself.
        """
        count = 1       # the Root
        for node in self._slots:
            if node and isinstance(node, Table):
                count += node.table_count
        return count

    def delete_leaf(self, key):
        """ Delete a Leaf node in or below this Root, given its key. """

        hcode = uhash(key)
        ndx = hcode & self._mask
        node = self._slots[ndx]
#       # DEBUG
#       print("Root.delete_leaf: ndx %d" % ndx)
#       if not node:
#           print("  slot is empty")
#       elif isinstance(node, Leaf):
#           print("  slot is a Leaf")
#       elif isinstance(node, Table):
#           print("  slot is a Table")
#       else:
#           print("  DUNNO what's in this slot")
#       # END
        if node is None:
            raise HamtNotFound
        if isinstance(node, Leaf):
            if node.key == key:
                # DEBUG
                # print("  FOUND: setting slot %d to None" % ndx)
                # END
                self._slots[ndx] = None
            else:
                raise HamtNotFound
        else:
            # entry is a Table, so recurse
            # DEBUG
            # print("  slot %d holds a Table" % ndx)
            # END
            if self._max_table_depth < 1:
                raise HamtNotFound
            node.delete_leaf(hcode >> self._texp, key)

    def find_leaf(self, key):
        """
        Find a Leaf entry given its key, searching from the Root.

        Given a properly shifted hashcode and the key for an entry,
        return the value associated with the entry or None if there
        is no such entry.
        """

        value = None
        hcode = uhash(key)
        ndx = hcode & self._mask
        node = self._slots[ndx]

        # DEBUG
#       print("Root.find_leaf:")
#       print("    hcode    0x%x" % hcode)
#       print("    ndx      0x%x (%d)" % (ndx, ndx))

        # END
        if node:
            if isinstance(node, Leaf):
                if node.key == key:
                    value = node.value
            else:
                if self._max_table_depth > 0:
                    # it's a Table, so recurse
                    value = node.find_leaf(hcode >> self.texp, 1, key)
        return value

    def insert_leaf(self, leaf):
        """ Insert a Leaf into or below the Root. """

        hcode = uhash(leaf.key)
        ndx = hcode & self._mask        # slot number
        # DEBUG
        # print("insert_leaf: hcode 0x%x" % hcode)
        # print("             mask  0x%x" % self._mask)
        # print("             ndx   0x%x (%d)" % (ndx, ndx))
        # END
        node = self._slots[ndx]

        if node is None:
            # the slot is empty
            self._slots[ndx] = leaf
        else:
            # there is something already in the slot
            if isinstance(node, Leaf):
                # it's a leaf; replace value iff the keys match
                cur_key = node.key
                new_key = leaf.key
                if cur_key == new_key:
                    # keys match
                    # DEBUG
                    # print("Root.insert_leaf: keys match, so overwriting val")
                    # END
                    node.value = leaf.value
                else:
                    # keys differ, so we replace node with a Table
                    if self._max_table_depth < 1:
                        raise HamtError(
                            "max table depth (%d) exceeded" %
                            self._max_table_depth)
                    new_hcode = hcode >> self._texp    # hcode for new entry

                    # DEBUG
#                   print(
#                       "Root.insert_leaf: keys differ, replacing with Table")
                    # END

                    new_table = Table(1, self, node)
                    self._slots[ndx] = new_table
                    new_table.insert_leaf(uhash(leaf.key) >> self._texp, leaf)

            else:
                # DEBUG
                #               print("Root.insert_leaf: node is Table")
                # END
                # it's a table
                if self._max_table_depth < 1:
                    raise HamtError(
                        "max table depth (%d) exceeded" %
                        self._max_table_depth)
                new_hcode = hcode >> self._texp    # hcode for new entry
                node.insert_leaf(new_hcode, leaf)

    # THIS CODE IS NEVER USED
#   def accept(self, func):
#       """ EXPERIMENT """
#       func(self)
